test/test_cpp_api_parity.py

```diff
@@ -1,687 +1,54 @@
-                    module_metadata=module_metadata,
-                    device=device,
-                    is_criterion=is_criterion)
-                test_name = 'test_torch_nn_{}'.format(test_params.module_variant_name)
-                torch_nn_test_params_map[test_name] = test_params
-
-                def test_fn(self):
-                    self._test_torch_nn_module_variant(test_params=torch_nn_test_params_map[self._testMethodName])
-
-                if device == 'cuda':
-                    test_fn = unittest.skipIf(not TEST_CUDA, "CUDA unavailable")(test_fn)
-
-                if not has_impl_parity:
-                    test_fn = unittest.expectedFailure(test_fn)
-
-                add_test(test_name, test_fn)
-
-        add_ctor_args_test_for_module(module_name, has_impl_parity)
-        add_variant_test_for_module(module_name, test_params_dict, has_impl_parity)
-
-add_torch_nn_module_tests(
-    sample_module.module_tests + common_nn.module_tests + common_nn.new_module_tests,
-    is_criterion=False)
-
-add_torch_nn_module_tests(
-    common_nn.criterion_tests + common_nn.new_criterion_tests,
-    is_criterion=True)
-
-# Assert that there exists auto-generated tests for SampleModule.
+class TestCppApiParity(common.TestCase):
+    pass
+
+expected_test_params_dicts = []
+
+for test_params_dicts, test_instance_class in [
+    (sample_module.module_tests, common_nn.ModuleTest),
+    (sample_functional.functional_tests, common_nn.NewModuleTest),
+    (common_nn.module_tests, common_nn.ModuleTest),
+    (common_nn.new_module_tests, common_nn.NewModuleTest),
+    (common_nn.criterion_tests, common_nn.CriterionTest),
+    (common_nn.new_criterion_tests, common_nn.NewCriterionTest),
+]:
+    module_impl_check.write_tests_to_test_class(
+        TestCppApiParity, test_params_dicts, test_instance_class, parity_table, devices)
+    functional_impl_check.write_tests_to_test_class(
+        TestCppApiParity, test_params_dicts, test_instance_class, parity_table, devices)
+    expected_test_params_dicts += test_params_dicts
+
+# Assert that all NN module/functional test dicts appear in the parity test
+assert len([name for name in TestCppApiParity.__dict__ if 'test_torch_nn_' in name]) == \
+    len(expected_test_params_dicts) * len(devices)
+
+# Assert that there exists auto-generated tests for `SampleModule` and `sample_functional`.
 assert len([name for name in TestCppApiParity.__dict__ if 'SampleModule' in name]) == \
-    len(sample_module.module_tests) * len(devices) + 1
+    len(sample_module.module_tests) * len(devices)
+assert len([name for name in TestCppApiParity.__dict__ if 'sample_functional' in name]) == \
+    len(sample_functional.functional_tests) * len(devices)
 
 
 if __name__ == "__main__":
+    module_impl_check.build_cpp_tests(TestCppApiParity, print_cpp_source=PRINT_CPP_SOURCE)
+    functional_impl_check.build_cpp_tests(TestCppApiParity, print_cpp_source=PRINT_CPP_SOURCE)
```

Comment:
TestCppApiParity doesn't work with pytest. I'm not sure how important it is that that works.

Reply:
I moved the two `build_cpp_tests` lines out from the `if __name__ == "__main__"` block, which makes `TestCppApiParity` work with pytest as well :D